File: attention-web1/src/types/calibration.ts
export type CalibrationPointId =
  | "center"
  | "topLeft"
  | "topRight"
  | "bottomLeft"
  | "bottomRight";

export interface CalibrationPointDefinition {
  id: CalibrationPointId;
  label: string;
  instruction: string;
  xPercent: number;
  yPercent: number;
}

export interface CalibrationPointCapture {
  id: CalibrationPointId;
  label: string;
  xPercent: number;
  yPercent: number;
  startMs: number;
  endMs: number;
}

export interface SessionCalibration {
  version: 1;
  source: "default" | "session";
  skipped: boolean;
  completed: boolean;
  capturedAt: string;
  pointCaptures: CalibrationPointCapture[];
  /** Ms from recording start when stimulus visuals begin (analysis trim point). */
  stimulusStartMs?: number;
  analysisStartMs?: number;
}

/** Default gaze baseline used for uploads and skipped calibration. */
export const DEFAULT_CALIBRATION_BASELINE = {
  hRatioCenter: 0.5,
  vRatioCenter: 0.5,
  yawNeutral: 0,
  pitchNeutral: 0,
  irisLeftThreshold: 0.38,
  irisRightThreshold: 0.62,
  irisUpThreshold: 0.38,
  irisDownThreshold: 0.62,
  yawLeftThreshold: -15,
  yawRightThreshold: 15,
  pitchUpThreshold: -20,
  pitchDownThreshold: 20,
} as const;

export const CALIBRATION_POINTS: CalibrationPointDefinition[] = [
  {
    id: "center",
    label: "Center",
    instruction: "Look at the center dot",
    xPercent: 50,
    yPercent: 50,
  },
  {
    id: "topLeft",
    label: "Top left",
    instruction: "Look at the top-left dot",
    xPercent: 8,
    yPercent: 8,
  },
  {
    id: "topRight",
    label: "Top right",
    instruction: "Look at the top-right dot",
    xPercent: 92,
    yPercent: 8,
  },
  {
    id: "bottomLeft",
    label: "Bottom left",
    instruction: "Look at the bottom-left dot",
    xPercent: 8,
    yPercent: 92,
  },
  {
    id: "bottomRight",
    label: "Bottom right",
    instruction: "Look at the bottom-right dot",
    xPercent: 92,
    yPercent: 92,
  },
];

export const CALIBRATION_HOLD_MS = 2500;

export function createDefaultSessionCalibration(): SessionCalibration {
  return {
    version: 1,
    source: "default",
    skipped: true,
    completed: false,
    capturedAt: new Date().toISOString(),
    pointCaptures: [],
  };
}

export function createSessionCalibrationFromCaptures(
  pointCaptures: CalibrationPointCapture[],
  skipped: boolean,
  stimulusStartMs?: number
): SessionCalibration {
  const startMs = stimulusStartMs ?? 0;
  return {
    version: 1,
    source: skipped ? "default" : "session",
    skipped,
    completed: !skipped && pointCaptures.length === CALIBRATION_POINTS.length,
    capturedAt: new Date().toISOString(),
    pointCaptures,
    stimulusStartMs: startMs,
    analysisStartMs: startMs,
  };
}

export function withStimulusStart(
  calibration: SessionCalibration,
  stimulusStartMs: number
): SessionCalibration {
  return {
    ...calibration,
    stimulusStartMs,
    analysisStartMs: stimulusStartMs,
  };
}

File: backend/pipeline/calibration.py
"""Per-session gaze calibration profiles."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

DEFAULT_CALIBRATION: dict[str, Any] = {
    "version": 1,
    "source": "default",
    "skipped": True,
    "completed": False,
    "baseline": {
        "hRatioCenter": 0.5,
        "vRatioCenter": 0.5,
        "yawNeutral": 0.0,
        "pitchNeutral": 0.0,
        "irisLeftThreshold": 0.38,
        "irisRightThreshold": 0.62,
        "irisUpThreshold": 0.38,
        "irisDownThreshold": 0.62,
        "yawLeftThreshold": -15.0,
        "yawRightThreshold": 15.0,
        "pitchUpThreshold": -20.0,
        "pitchDownThreshold": 20.0,
    },
    "pointCaptures": [],
}


def analysis_start_seconds(calibration: dict[str, Any]) -> float:
    """Seconds into the raw recording where stimulus / analysis begins."""
    raw_ms = calibration.get("analysisStartMs", calibration.get("stimulusStartMs", 0))
    try:
        return max(0.0, float(raw_ms) / 1000.0)
    except (TypeError, ValueError):
        return 0.0


def load_calibration(work_dir: Path) -> dict[str, Any]:
    calibration_path = work_dir / "calibration.json"
    if not calibration_path.exists():
        return dict(DEFAULT_CALIBRATION)

    try:
        payload = json.loads(calibration_path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return dict(DEFAULT_CALIBRATION)

    if not isinstance(payload, dict):
        return dict(DEFAULT_CALIBRATION)
    return payload


def save_calibration(work_dir: Path, payload: dict[str, Any] | str | None) -> dict[str, Any]:
    work_dir.mkdir(parents=True, exist_ok=True)
    if not payload:
        calibration = dict(DEFAULT_CALIBRATION)
    elif isinstance(payload, str):
        try:
            parsed = json.loads(payload)
        except json.JSONDecodeError:
            parsed = {}
        calibration = parsed if isinstance(parsed, dict) else dict(DEFAULT_CALIBRATION)
    else:
        calibration = payload

    if "baseline" not in calibration:
        calibration["baseline"] = dict(DEFAULT_CALIBRATION["baseline"])

    (work_dir / "calibration.json").write_text(
        json.dumps(calibration, indent=2),
        encoding="utf-8",
    )
    return calibration
